### PR TITLE
Fixed #402.

### LINKED ISSUE #402: Syntax error is not raised when a wrong command is passed to the shell
CPython 3.7:
```shell
>>> printtype.__doc__)
  File "<stdin>", line 1
    printtype.__doc__)
                     ^
SyntaxError: invalid syntax
```

RustPython:
```shell
>>>>> printtype.__doc__)
thread 'main' panicked at 'attempt to subtract with overflow', parser/src/lexer.rs:907:21
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

### PR DESCRIPTION
**Pull request by Tom1380**

### CODE REVIEW

**Comment by cthulahoops** [comment]
Change looks good. Could you describe the change in the commit message? "Better error message on unexpected ')'.", perhaps?

This doesn't fix 402.

**Comment by Tom1380** [comment]
@cthulahoops I'm not sure how to return a Syntax Error as I'm new to the project, but if I'm correct we just need to replace the panic with a line of code raising it. As I said I don't know how to, can you help me? After that it should be fixed.

**Comment by cthulahoops** [comment]
I can do, but I need to go out now. I'll have a look later on?

**Comment by Tom1380** [comment]
@cthulahoops yes, that would be great, thank you :+1:.

**Comment by Tom1380** [comment]
I managed to make it work. Rustfmt complained about the indentation on Travis, but I'm not sure what the problem was judging by the output, so I ran rustfmt to format it automatically and pushed a new commit to prevent rustfmt from complaining. @cthulahoops

**Comment by cthulahoops** [comment]
Ah, looks like it's easier than I expected.

Adding another error type to the lexical error class:

```
#[derive(Debug)]
pub enum LexicalError {
    StringError,
    NestingError,
}
```

```
                    if self.nesting == 0 {
                        return Some(Err(LexicalError::NestingError));
                    }
```

And then it seems to work.

File: parser/src/lexer.rs [hunk 1]
```diff
@@ -914,7 +918,9 @@ where
                 }
                 Some(']') => {
                     let result = self.eat_single_char(Tok::Rsqb);
-                    self.nesting -= 1;
+                    if self.closing_bracket_is_valid() {
```
Comment (cthulahoops) [inline parser/src/lexer.rs]: I wouldn't use this function. For me it's less clear and longer than the code it replaces.

Comment (Tom1380) [reply parser/src/lexer.rs]: Ok, what should I do? Should I create a new pull request from scratch or simply make a new commit?

Comment (cthulahoops) [reply parser/src/lexer.rs]: I'd continue to push to this branch. Once it's ready to go we can rebase to clean up. (I presume you can force push here, not sure how github permissions work.)

Comment (Tom1380) [reply parser/src/lexer.rs]: Ok I pushed the commit, what should I do now to clean the history without  messing everything up?

Comment (cthulahoops) [reply parser/src/lexer.rs]: It looks like under the circumstances I can use github squash-merge to achieve something similar, so I'll just do that and you don't need to do anything.

Thanks for your work on this. Looking good now.

My approach would be:

```
git rebase -i HEAD~6
```

(I'd normally use ```git rebase -i $(git merge-base HEAD origin/master)```, but I think origin/master will be the same as master in your case, so I'm just making it easy and saying the last six commits.)

Change first commit to reword, all subsequent ones to fixup, then:

```
git push -f
```

Comment (cthulahoops) [reply parser/src/lexer.rs]: And I typo'd the commit message. :(

**Comment by Tom1380** [comment]
s@cthulahoops Thank you a lot for the help, this is the first time I make a pull request :+1:.

**Pull request merged**